src/app/users/get-user.ts

```diff
@@ -1,65 +1,99 @@
-import {
-  SemanticAttributes,
-  asyncRunInSpan,
-  addAttributesToCurrentSpan,
-  ENDUSER_ALIAS,
-} from "@services/tracing"
+import { SemanticAttributes, asyncRunInSpan } from "@services/tracing"
 import { UsersRepository } from "@services/mongoose"
-import { IpFetcher } from "@services/ipfetcher"
 import { getIpConfig } from "@config/app"
+import { IpFetcher } from "@services/ipfetcher"
+import { UsersIpRepository } from "@services/mongoose/users-ip"
 import { RepositoryError } from "@domain/errors"
 
 const users = UsersRepository()
+const usersIp = UsersIpRepository()
 
 export const getUserForLogin = async ({
   userId,
   ip,
+  logger,
 }: {
-  userId: string
-  ip?: string
+  userId: UserId
+  ip?: Ip
+  logger: Logger
 }): Promise<User | ApplicationError> =>
   asyncRunInSpan(
     "app.getUserForLogin",
     { [SemanticAttributes.CODE_FUNCTION]: "getUserForLogin" },
     async () => {
-      const user = await users.findById(userId as UserId)
+      const user = await users.findById(userId)
+
       if (user instanceof Error) {
         return user
       }
-      addAttributesToCurrentSpan({
-        [ENDUSER_ALIAS]: user.username,
-      })
-      user.lastConnection = new Date()
 
-      // IP tracking logic could be extracted into domain
+      // this routing run asynchrously, to update metadata on the background
+      updateUserIpInfo({ userId, ip, logger })
+
+      return user
+    },
+  )
+
+const updateUserIpInfo = async ({
+  userId,
+  ip,
+  logger,
+}: {
+  userId: UserId
+  ip?: Ip
+  logger: Logger
+}): Promise<void> =>
+  asyncRunInSpan(
+    "app.addIp",
+    { [SemanticAttributes.CODE_FUNCTION]: "addIp" },
+    async () => {
       const ipConfig = getIpConfig()
-      if (ip && ipConfig.ipRecordingEnabled) {
-        const lastIP: IPType | undefined = user.lastIPs.find(
-          (ipObject: IPType) => ipObject.ip === ip,
-        )
-        if (lastIP) {
-          lastIP.lastConnection = user.lastConnection
-        } else if (ipConfig.proxyCheckingEnabled) {
+
+      const lastConnection = new Date()
+
+      const userIp = await usersIp.findById(userId)
+
+      if (userIp instanceof RepositoryError) throw userIp
+
+      if (!ip || !ipConfig.ipRecordingEnabled) {
+        const result = await usersIp.update(userId, lastConnection)
+
+        if (result instanceof Error) {
+          logger.error(
+            { result, userId, ip },
+            "impossible to update user last connection",
+          )
+        }
+
+        return
```

Comment:
Why aren't you returning the error here? Returning it will mean it automatically gets reported via OpenTelemetry `span.recordException` in the wrapper function `runInSpan`... eating it means we loose that information in honeycomb.

Reply:
I didn't returned it because this is a promise that is not being awaited and there will be not action done whether this promise succeed or not.

if I'm return (or throwing?) an error this would have an effect on honeycomb?

Reply:
Errors should always be returned from app layer. Its up to the client what to do with it (including silently ignore).

Apart from that general approach to errors there is also the added benefit that since you are wrapping the function with `runInSpan` it will automatically report the error to honeycomb - but regardless you should always return it and let the client decide to ignore or not.

Also IMO logging should happen at a higher level where we decide if the error is unrecoverable or returned to user or just eaten etc. That decision should happen at the top of the call stack - outside of app layer and that's where we should log the decision IMO. Passing around the logger down into the call stack is a lot of boilerplate without much added value. But its not a big deal - as some of our monitoring relies on logging we haven't completely cleaned it up yet.

Reply:
> but regardless you should always return it and let the client decide to ignore or not.

maybe you miss my point above: we're not bubbling up this function

Reply:
> Also IMO logging should happen at a higher level where we decide if the error is unrecoverable or returned to user or just eaten etc. That decision should happen at the top of the call stack - outside of app layer and that's where we should log the decision IMO. Passing around the logger down into the call stack is a lot of boilerplate without much added value. But its not a big deal - as some of our monitoring relies on logging we haven't completely cleaned it up yet.

your assumption is that you only want to log the critical issue. I guess my assumption is that we can log more lower level call (typically called "info", "debug" or "trace")

Reply:
> maybe you miss my point above: we're not bubbling up this function

Ah do you mean because its not `export`ed there is no external client? In many languages (ie rust, Golang) its still common to return the error unless it's a top level function (like `main`) so it's just something I consider normal. I guess its not idiomatic in JS due to `try/catch` semantics but still it would be good here so that honeycomb will receive it.

> your assumption is that you only want to log the critical issue. I guess my assumption is that we can log more lower level call (typically called "info", "debug" or "trace")

Yeah I have no problem with that. We'll see how it evolves with the usage of the OTEL libraries - how we want to approach logging vs tracing etc in the future.

Reply:
btw, we should consider honeycomb as a complement: we can't rely on this for our tech stack as the sole way to debug our application because this is not FOSS. so saying "this works with honeycomb" is not something we should emphasize if we/other people using the stack don't have another solution to debug issues

Reply:
> Ah do you mean because its not exported there is no external client

no because this function is called as a "fire and forget". this is the whole point of this whole PR: run this function asynchronously from the rest of the graphql flow.

Reply:
OTEL is open source - if we don't want to rely on honeycomb we can deploy a jaeger UI. We have nothing honeycomb specific in our codebase. The only thing related to honeycomb is where our central `opentelemetry-collector` forwards the traces for visualising. All comments I said above regarding returning errors you can replace the name `honeycomb` with `opentelemetry-collector`.

Reply:
Any tracing visualiser can be used to help debugging. That we ship to honeycomb is an implementation detail, but not of our app - rather of how we've configured the collector.

If you don't want to return the error than you can also do the following:
```
const span = trace.getSpan(context.active())
span.recordException(error)
return
```

This is the equivalent code like adding the log.

But this is boilerplate you can avoid because of https://github.com/GaloyMoney/galoy/blob/main/src/services/tracing.ts#L159-L162
This has nothing to do with honeycomb.